File: src/test/java/com/github/jactor/persistence/dto/PersonInternalDtoTest.java
package com.github.jactor.persistence.dto;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.time.LocalDateTime;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

@DisplayName("A PersonDto")
class PersonInternalDtoTest {

    @DisplayName("should have a copy constructor")
    @Test void shouldHaveCopyConstructor() {
        PersonInternalDto personInternalDto = new PersonInternalDto();
        personInternalDto.setAddress(new AddressInternalDto());
        personInternalDto.setDescription("description");
        personInternalDto.setFirstName("first name");
        personInternalDto.setLocale("no");
        personInternalDto.setSurname("surname");

        PersonInternalDto copied = new PersonInternalDto(personInternalDto.getPersistentDto(), personInternalDto);

        assertAll(
                () -> assertThat(copied.getAddress()).as("address").isEqualTo(personInternalDto.getAddress()),
                () -> assertThat(copied.getDescription()).as("description").isEqualTo(personInternalDto.getDescription()),
                () -> assertThat(copied.getFirstName()).as("first name").isEqualTo(personInternalDto.getFirstName()),
                () -> assertThat(copied.getLocale()).as("locale").isEqualTo(personInternalDto.getLocale()),
                () -> assertThat(copied.getSurname()).as("surname").isEqualTo(personInternalDto.getSurname())
        );
    }

    @DisplayName("should give values to PersistentDto")
    @Test void shouldGiveValuesToPersistentDto() {
        PersistentDto persistentDto = new PersistentDto();
        persistentDto.setCreatedBy("jactor");
        persistentDto.setTimeOfCreation(LocalDateTime.now());
        persistentDto.setId(1L);
        persistentDto.setModifiedBy("tip");
        persistentDto.setTimeOfModification(LocalDateTime.now());

        PersistentDto copied = new PersonInternalDto(persistentDto, new PersonInternalDto()).getPersistentDto();

        assertAll(
                () -> assertThat(copied.getCreatedBy()).as("created by").isEqualTo(persistentDto.getCreatedBy()),
                () -> assertThat(copied.getTimeOfCreation()).as("creation time").isEqualTo(persistentDto.getTimeOfCreation()),
                () -> assertThat(copied.getId()).as("id").isEqualTo(persistentDto.getId()),
                () -> assertThat(copied.getModifiedBy()).as("updated by").isEqualTo(persistentDto.getModifiedBy()),
                () -> assertThat(copied.getTimeOfModification()).as("updated time").isEqualTo(persistentDto.getTimeOfModification())
        );
    }
}

File: src/test/kotlin/com/github/jactor/persistence/dto/PersonInternalDtoTest.kt
package com.github.jactor.persistence.dto

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertAll
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import java.time.LocalDateTime

@DisplayName("A PersonDto")
internal class PersonInternalDtoTest {

    @Test
    fun `should have a copy constructor`() {
        val personInternalDto = PersonInternalDto()
        personInternalDto.address = AddressInternalDto()
        personInternalDto.description = "description"
        personInternalDto.firstName = "first name"
        personInternalDto.locale = "no"
        personInternalDto.surname = "surname"

        val (_, address, locale, firstName, surname, description) = PersonInternalDto(personInternalDto.persistentDto, personInternalDto)

        assertAll(
            { assertThat(address).`as`("address").isEqualTo(personInternalDto.address) },
            { assertThat(description).`as`("description").isEqualTo(personInternalDto.description) },
            { assertThat(firstName).`as`("first name").isEqualTo(personInternalDto.firstName) },
            { assertThat(locale).`as`("locale").isEqualTo(personInternalDto.locale) },
            { assertThat(surname).`as`("surname").isEqualTo(personInternalDto.surname) }
        )
    }

    @DisplayName("should give values to PersistentDto")
    @Test
    fun shouldGiveValuesToPersistentDto() {
        val persistentDto = PersistentDto()
        persistentDto.createdBy = "jactor"
        persistentDto.timeOfCreation = LocalDateTime.now()
        persistentDto.id = 1L
        persistentDto.modifiedBy = "tip"
        persistentDto.timeOfModification = LocalDateTime.now()

        val (id, createdBy, timeOfCreation, modifiedBy, timeOfModification) = PersonInternalDto(persistentDto, PersonInternalDto()).persistentDto

        assertAll(
            { assertThat(createdBy).`as`("created by").isEqualTo(persistentDto.createdBy) },
            { assertThat(timeOfCreation).`as`("creation time").isEqualTo(persistentDto.timeOfCreation) },
            { assertThat(id).`as`("id").isEqualTo(persistentDto.id) },
            { assertThat(modifiedBy).`as`("updated by").isEqualTo(persistentDto.modifiedBy) },
            { assertThat(timeOfModification).`as`("updated time").isEqualTo(persistentDto.timeOfModification) }
        )
    }
}
